cloud/edgecontroller/pkg/devicecontroller/controller/downstream.go

```diff
@@ -0,0 +1,742 @@
+			case watch.Deleted:
+				dc.deviceModelDeleted(deviceModel)
+			case watch.Modified:
+				dc.deviceModelUpdated(deviceModel)
+			default:
+				log.LOGGER.Warnf("deviceModel event type: %s unsupported", e.Type)
+			}
+		case <-stop:
+			log.LOGGER.Infof("stop syncDeviceModel")
+			running = false
+		}
+	}
+}
+
+func (dc *DownstreamController) deviceModelAdded(deviceModel *v1alpha1.DeviceModel) {
+	// nothing to do when deviceModel added, only add in map
+	dc.deviceModelManager.DeviceModel.Store(deviceModel.Name, &CacheDeviceModel{ObjectMeta: deviceModel.ObjectMeta, Spec: deviceModel.Spec})
+}
+
+func isDeviceModelUpdated(old *CacheDeviceModel, new *v1alpha1.DeviceModel) bool {
+	// does not care fields
+	old.ObjectMeta.ResourceVersion = new.ObjectMeta.ResourceVersion
+	old.ObjectMeta.Generation = new.ObjectMeta.Generation
+
+	// return true if ObjectMeta or Spec or Status changed, else false
+	return !reflect.DeepEqual(old.ObjectMeta, new.ObjectMeta) || !reflect.DeepEqual(old.Spec, new.Spec)
+}
+
+func (dc *DownstreamController) deviceModelUpdated(deviceModel *v1alpha1.DeviceModel) {
+	value, ok := dc.deviceModelManager.DeviceModel.Load(deviceModel.Name)
+	dc.deviceModelManager.DeviceModel.Store(deviceModel.Name, &CacheDeviceModel{ObjectMeta: deviceModel.ObjectMeta, Spec: deviceModel.Spec})
+	if ok {
+		cachedDeviceModel := value.(*CacheDeviceModel)
+		if isDeviceModelUpdated(cachedDeviceModel, deviceModel) {
+			dc.updateAllConfigMaps(deviceModel)
+		}
+	} else {
+		dc.deviceModelAdded(deviceModel)
+	}
+}
+
+func (dc *DownstreamController) updateAllConfigMaps(deviceModel *v1alpha1.DeviceModel) {
+	//TODO: add logic to update all config maps, How to manage if a property is deleted but a device is referring that property. Need to come up with a design.
+}
+
+func (dc *DownstreamController) deviceModelDeleted(deviceModel *v1alpha1.DeviceModel) {
+	// TODO: Need to use finalizer like method to delete all devices referring to this model. Need to come up with a design.
+	dc.deviceModelManager.DeviceModel.Delete(deviceModel.Name)
+}
+
+func (dc *DownstreamController) syncDevice(stop chan struct{}) {
+	running := true
+	for running {
+		select {
+		case e := <-dc.deviceManager.Events():
+			device, ok := e.Object.(*v1alpha1.Device)
+			if !ok {
+				log.LOGGER.Warnf("Object type: %T unsupported", device)
+				continue
+			}
+			switch e.Type {
+			case watch.Added:
+				dc.deviceAdded(device)
+			case watch.Deleted:
+				dc.deviceDeleted(device)
+			case watch.Modified:
+				dc.deviceUpdated(device)
+			default:
+				log.LOGGER.Warnf("Device event type: %s unsupported", e.Type)
+			}
+		case <-stop:
+			log.LOGGER.Infof("Stop syncDevice")
+			running = false
+		}
+	}
+}
+
+// addToConfigMap adds device in the configmap
+func (dc *DownstreamController) addToConfigMap(device *v1alpha1.Device) {
+	configMap, ok := dc.configMapManager.ConfigMap.Load(device.Spec.NodeSelector.NodeSelectorTerms[0].MatchExpressions[0].Values[0])
```

Comment:
we can store the node name from device.Spec.NodeSelector.NodeSelectorTerms[0].MatchExpressions[0].Values[0] in a local variable and then use the variable at all places below.
What if the node selector is empty ? i.e if  a device is not bound to any node yet ? 
What if the node selector has multiple values ? say I need device A to be bound to nodes A and B ?

Reply:
If device has no node selector, we only add in cache and ignore.
If device has 2 node selectors, we just take the first one. 
Will add this point in common issue and lets handle in next version.